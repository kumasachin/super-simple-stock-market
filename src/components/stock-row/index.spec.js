import { render, screen, fireEvent, cleanup } from "@testing-library/react";
import StockRow from "./";

describe("Stocks row snapshot testing", () => {
  afterEach(cleanup);
  const stock = {
    symbol: "AER",
    type: "Common",
    lastDividend: 123,
    parValue: 100,
    price: 100,
    quantity: 1
  };
  const trades = [{
    symbol: "AER",
    type: "Common",
    lastDividend: "-",
    fixedDividend: "23",
    parValue: "-",
    price: "123",
    quantity: "21"
  }];
  it("should take a snapshot", () => {
    const { asFragment } = render(<table><tbody><StockRow stock={stock} trades={trades} index={0} setStockInStore={jest.fn()} /></tbody></table>)
    expect(asFragment(<table><tbody><StockRow /></tbody></table>)).toMatchSnapshot()
  });

  it("should render row with valid data", async () => {
    render(<table><tbody><StockRow stock={stock} trades={trades} index={0} setStockInStore={jest.fn()} /></tbody></table>)
    expect(screen.getByTestId("stock-row-0")).toHaveTextContent("AERCommon123-100123.00BuySell")
  });
});

describe("Stocks row function validation", () => {
  afterEach(cleanup);
  const stock01 = {
    symbol: "AER",
    type: "Common",
    lastDividend: 123,
    fixedDividend: 2,
    parValue: 100,
    price: 100,
    quantity: 1
  };
  const trades = [{
    symbol: "AER",
    type: "Common",
    lastDividend: "-",
    fixedDividend: "23",
    parValue: "-",
    price: "123",
    quantity: "21"
  }];

  it("should validate stock rows with type common", () => {
    render(<table><tbody><StockRow stock={stock01} trades={trades}  index={0} setStockInStore={jest.fn()} handleTrade= {jest.fn()} /></tbody></table>)
    const buybtn = screen.getByTestId("stock-buy-0");
    const sellbtn = screen.getByTestId("stock-sell-0");
    expect(buybtn).toBeEnabled();
    expect(sellbtn).toBeEnabled();
    fireEvent.blur(screen.getByTestId("stock-price-0"), { target: { value: "223" }});
    fireEvent.change(screen.getByTestId("stock-quantity-0"), { target: { value: "2" }});
    expect(buybtn).toBeEnabled();
    expect(sellbtn).toBeEnabled();
    expect(screen.getByTestId("stock-type-0")).toHaveTextContent("Common");
    expect(screen.getByTestId("stock-quantity-0")).toHaveValue(2);
    expect(screen.getByTestId("stock-price-0")).toHaveValue(223);
    expect(screen.getByTestId("stock-symbol-0")).toBeTruthy();
    expect(screen.getByTestId("stock-symbol-0")).toHaveTextContent("AER");
    expect(screen.getByTestId("stock-lastdivident-0")).toHaveTextContent("2");
    expect(screen.getByTestId("stock-fixeddividend-0")).toHaveTextContent("2%");
    expect(screen.getByTestId("stock-parvalue-0")).toHaveTextContent("100");
    fireEvent.click(buybtn);
    expect(screen.getByTestId("stock-dividentyield-0")).toHaveTextContent("0.55");
    expect(screen.getByTestId("stock-peratio-0")).toHaveTextContent("404.30"); 
    expect(screen.getByTestId("stock-vwsprice-0")).toHaveTextContent("14.29"); 
  });

  it("should validate stock rows with type preferred", () => {
    const stock02 = {
      ...stock01,
      type: "Preferred"
    };
    render(<table><tbody><StockRow stock={stock02} trades={trades} index={0} setStockInStore={jest.fn()} handleTrade= {jest.fn()} /></tbody></table>)
    const buybtn = screen.getByTestId("stock-buy-0");
    const sellbtn = screen.getByTestId("stock-sell-0");
    expect(buybtn).toBeEnabled();
    expect(sellbtn).toBeEnabled();
    fireEvent.blur(screen.getByTestId("stock-price-0"), { target: { value: "223" }});
    fireEvent.change(screen.getByTestId("stock-quantity-0"), { target: { value: "2" }});
    expect(buybtn).toBeEnabled();
    expect(sellbtn).toBeEnabled();
    expect(screen.getByTestId("stock-type-0")).toHaveTextContent("Preferred");
    expect(screen.getByTestId("stock-quantity-0")).toHaveValue(2);
    expect(screen.getByTestId("stock-price-0")).toHaveValue(223);
    expect(screen.getByTestId("stock-symbol-0")).toBeTruthy();
    expect(screen.getByTestId("stock-symbol-0")).toHaveTextContent("AER");
    expect(screen.getByTestId("stock-lastdivident-0")).toHaveTextContent("2");
    expect(screen.getByTestId("stock-fixeddividend-0")).toHaveTextContent("2%");
    expect(screen.getByTestId("stock-parvalue-0")).toHaveTextContent("100");
    fireEvent.click(buybtn);
    expect(screen.getByTestId("stock-dividentyield-0")).toHaveTextContent("0.90");
    expect(screen.getByTestId("stock-peratio-0")).toHaveTextContent("248.64"); 
    expect(screen.getByTestId("stock-vwsprice-0")).toHaveTextContent("14.29"); 
  });

  it("should validate stock rows with type preferred", () => {
    const stock03 = {
      symbol: "AER",
      type: "Common",
      parValue: 100,
      price: 100,
      quantity: 1
    };
    render(<table><tbody><StockRow stock={stock03} trades={trades} index={0} setStockInStore={jest.fn()} handleTrade= {jest.fn()} /></tbody></table>)
    const buybtn = screen.getByTestId("stock-buy-0");
    const sellbtn = screen.getByTestId("stock-sell-0");
    expect(buybtn).toBeEnabled();
    expect(sellbtn).toBeEnabled();
    fireEvent.blur(screen.getByTestId("stock-price-0"), { target: { value: "223" }});
    fireEvent.change(screen.getByTestId("stock-quantity-0"), { target: { value: "2" }});
    expect(buybtn).toBeEnabled();
    expect(sellbtn).toBeEnabled();
    expect(screen.getByTestId("stock-type-0")).toHaveTextContent("Common");
    expect(screen.getByTestId("stock-quantity-0")).toHaveValue(2);
    expect(screen.getByTestId("stock-price-0")).toHaveValue(223);
    expect(screen.getByTestId("stock-symbol-0")).toBeTruthy();
    expect(screen.getByTestId("stock-symbol-0")).toHaveTextContent("AER");
    expect(screen.getByTestId("stock-lastdivident-0")).toHaveTextContent("-");
    expect(screen.getByTestId("stock-fixeddividend-0")).toHaveTextContent("-");
    expect(screen.getByTestId("stock-parvalue-0")).toHaveTextContent("100");
    fireEvent.click(buybtn);
    expect(screen.getByTestId("stock-dividentyield-0")).toHaveTextContent("");
    expect(screen.getByTestId("stock-peratio-0")).toHaveTextContent(""); 
    expect(screen.getByTestId("stock-vwsprice-0")).toHaveTextContent("14.29"); 
  });
});

