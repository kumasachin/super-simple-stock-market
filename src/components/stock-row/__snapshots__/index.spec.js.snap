// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stocks row snapshot testing should take a snapshot 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr
        data-testid="stock-row-0"
      >
        <td
          data-testid="stock-symbol-0"
        >
          AER
        </td>
        <td
          data-testid="stock-type-0"
        >
          Common
        </td>
        <td
          data-testid="stock-lastdivident-0"
        >
          123
        </td>
        <td
          data-testid="stock-fixeddividend-0"
        >
          -
        </td>
        <td
          data-testid="stock-parvalue-0"
        >
          100
        </td>
        <td>
          <input
            data-testid="stock-price-0"
            min="1"
            name="price"
            type="number"
          />
        </td>
        <td
          data-testid="stock-dividentyield-0"
        />
        <td
          data-testid="stock-peratio-0"
        />
        <td
          data-testid="stock-vwsprice-0"
        >
          123.00
        </td>
        <td>
          <input
            data-testid="stock-quantity-0"
            min="1"
            name="quantity"
            type="number"
            value="1"
          />
        </td>
        <td
          data-testid="stock-buysell-0"
        >
          <button
            data-testid="stock-buy-0"
            name="buy"
          >
            Buy
          </button>
          <br />
          <br />
          <button
            data-testid="stock-sell-0"
            name="sell"
          >
            Sell
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`Stocks row snapshot testing should take a snapshot 2`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr
        data-testid="stock-row-0"
      >
        <td
          data-testid="stock-symbol-0"
        >
          AER
        </td>
        <td
          data-testid="stock-type-0"
        >
          Common
        </td>
        <td
          data-testid="stock-lastdivident-0"
        >
          12
        </td>
        <td
          data-testid="stock-fixeddividend-0"
        >
          2%
        </td>
        <td
          data-testid="stock-parvalue-0"
        >
          100
        </td>
        <td>
          <input
            data-testid="stock-price-0"
            min="1"
            name="price"
            type="number"
          />
        </td>
        <td
          data-testid="stock-dividentyield-0"
        />
        <td
          data-testid="stock-peratio-0"
        />
        <td
          data-testid="stock-vwsprice-0"
        >
          123.00
        </td>
        <td>
          <input
            data-testid="stock-quantity-0"
            min="1"
            name="quantity"
            type="number"
            value=""
          />
        </td>
        <td
          data-testid="stock-buysell-0"
        >
          <button
            data-testid="stock-buy-0"
            disabled=""
            name="buy"
          >
            Buy
          </button>
          <br />
          <br />
          <button
            data-testid="stock-sell-0"
            disabled=""
            name="sell"
          >
            Sell
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
